gen-certs.sh

```diff
@@ -0,0 +1,147 @@
+
+    mkdir -p $(work)
+
+    # generate the CA _work key
+    openssl genrsa -out $(dir)/bootstrap-ca.key 4096
+
+    cat > $(work)/ca.cfg <<EOF
+[ca]
+default_ca = CA_default
+
+[CA_default]
+dir = $(dir)
+certs	= \$dir
+crl_dir	= $(work)
+database = $(work)/index.txt
+new_certs_dir	= \$dir
+
+certificate	= \$dir/
+serial = $(work)/serial
+crlnumber	= $(work)/crlnumber
+crl = \$dir/crl.pem
+private_key	= \$dir/bootstrap-ca.key
+RANDFILE = \$dir/.rand
+
+default_days = 365
+default_crl_days = 30
+default_md = default
+preserve = false
+copy_extensions = copy
+
+policy          = policy_match
+
+[ policy_match ]
+countryName             = match
+stateOrProvinceName     = match
+organizationName        = match
+organizationalUnitName  = optional
+commonName              = supplied
+emailAddress            = optional
+
+[req]
+distinguished_name = req_distinguished_name
+req_extensions = v3_req
+prompt = no
+
+[req_distinguished_name]
+C = $COUNTRY
+ST = $STATE
+L = $CITY
+O = $ORG
+OU = $ORGUNIT
+CN = $CACN
+
+[v3_req]
+# Extensions to add to a certificate request
+basicConstraints = CA:TRUE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+EOF
+
+    rm -f $(work)/index.txt $(work)/index.txt.attr
+    touch $(work)/index.txt $(work)/index.txt.attr
+    echo 1000 > $(work)/serial > $(work)/crlnumber
+
+    openssl req -batch -config $(work)/ca.cfg -sha256 -new -x509 -days 3650 -key $(dir)/bootstrap-ca.key -out $(dir)/bootstrap-ca.crt
+}
+
+gencert() {
+    genca
+
+    [ -f $(dir)/$1.key ] && [ -f $(dir)/$1.crt ] && return
+
+    # generate the server cert
+    openssl genrsa -out $(dir)/$1.key 4096
+
+    cat > $(work)/$1.cfg <<EOF
+[req]
+distinguished_name = req_distinguished_name
+req_extensions = v3_req
+prompt = no
+
+[req_distinguished_name]
+C = $COUNTRY
+ST = $STATE
+L = $CITY
+O = $ORG
+OU = $ORGUNIT
+CN = $2
+
+[v3_req]
+# Extensions to add to a certificate request
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+subjectAltName = @alt_names
+
+[alt_names]
+EOF
+
+    count=0
+    for dnsalt in $3
+    do
+        count=$((count + 1))
+        echo "DNS.${count} = ${dnsalt}" >> $(work)/$1.cfg
+    done
+
+    count=0
+    for ipalt in $4
+    do
+        count=$((count + 1))
+        echo "IP.${count} = ${ipalt}" >> $(work)/$1.cfg
+    done
+
+    # generate the server csr
+    openssl req -batch -config $(work)/$1.cfg -new -sha256 -nodes -extensions v3_req -key $1.key -out $1.csr
+
+    # sign the server cert
+    openssl ca -batch -config $(work)/ca.cfg -extensions v3_req -notext -in $1.csr -out $1.crt
+
+    # final verification
+    openssl verify -CAfile $(dir)/bootstrap-ca.crt $1.crt
+}
+
+all_hostnames=$(echo "$(hostname) $(hostname --fqdn) $(hostnamectl --transient) $(hostnamectl --static) \
+                      $(cat /etc/hostname)" | tr ' ' '\n' | sort -u | tr '\n' ' ')
+
+gencert "velum" "$(hostname)" "$all_hostnames" "$(hostname --all-ip-addresses)"
```

Comment:
We should include all non-127.0.0.1 IPs, as that's quite likely the way people will connect in many environments  - e.g. OpenStack hostnames are meaningless.

For OpenStack and AWS, we could also include `curl http://169.254.169.254/latest/meta-data/public-ipv4` which is the floating IP, if any, assigned to a VM. I'd need to check what the return code/value is when no FIP is assigned though)